server/validation/validation.go

```diff
@@ -1,47 +1,313 @@
 package validation
 
 import (
-	"context"
 	"errors"
+	"fmt"
+	"reflect"
 
 	"github.com/openfga/openfga/pkg/tuple"
+	"github.com/openfga/openfga/pkg/typesystem"
 	"github.com/openfga/openfga/storage"
 	openfgapb "go.buf.build/openfga/go/openfga/api/openfga/v1"
 )
 
-// ValidateTuple checks whether a tuple is valid. If it is not, returns as error.
-// If it is, it grabs the type of the tuple's object, and the tuple's relation, and returns its corresponding userset.
-func ValidateTuple(ctx context.Context, backend storage.TypeDefinitionReadBackend, store, authorizationModelID string, tk *openfgapb.TupleKey) (*openfgapb.Userset, error) {
-	if err := tuple.ValidateUser(tk); err != nil {
-		return nil, err
+// ValidateTuple checks whether a tuple is welformed and valid according to the provided model.
+func ValidateTuple(model *openfgapb.AuthorizationModel, tk *openfgapb.TupleKey) error {
```

Comment:
I like taking the `*openfgapb.AuthorizationModel` but then we subsequently construct a typesystem off of it all over the place. I'm beginning to lean toward taking a `*typesystem.TypeSystem` instead. 

@miparnisari @craigpastro do you have a preference/opinion?

Reply:
I think taking a `TypeSystem` makes more sense.

Reply:
Refactored to take `TypeSystem`. That is better 👍